src/typechecker.erl

```diff
@@ -2038,68 +2202,243 @@ do_type_check_expr_in(Env, ResTy, {'try', _, Block, CaseCs, CatchCs, AfterBlock}
+            {VarBinds2, Cs2} = type_check_expr_in(EnvArg2, ResTy, Arg2),
+            {union_var_binds(VarBinds1, VarBinds2, Env#env.tenv),
+             constraints:combine([Cs, Cs1, Cs2])};
         false ->
-          throw({type_error, int_error, Op, P, ResTy})
-    end.
+            throw({type_error, logic_error, Op, P, ResTy})
+    end;
 type_check_logic_op_in(Env, ResTy, Op, P, Arg1, Arg2) ->
-    case subtype(ResTy, {type, erl_anno:new(0), 'bool', []}, Env#env.tenv) of
+    Bool = type(boolean),
+    case subtype(Bool, ResTy, Env#env.tenv) of
         {true, Cs} ->
-          {VarBinds1, Cs1} = type_check_expr_in(Env, ResTy, Arg1),
-          EnvArg2 =
-              if Op =:= 'andalso'; Op =:= 'orelse' ->
-                      %% variable bindings are propagated from Arg1 to Arg2
-                      Env#env{venv = union_var_binds([Env#env.venv, VarBinds1])};
-                 true ->
-                      Env
-              end,
-          {VarBinds2, Cs2} = type_check_expr_in(EnvArg2, ResTy, Arg2),
-          {union_var_binds([VarBinds1, VarBinds2])
+          {VarBinds1, Cs1} = type_check_expr_in(Env, Bool, Arg1),
+          {VarBinds2, Cs2} = type_check_expr_in(Env, Bool, Arg2),
+          {union_var_binds(VarBinds1, VarBinds2, Env#env.tenv)
           ,constraints:combine([Cs, Cs1, Cs2])};
         false ->
           throw({type_error, logic_error, Op, P, ResTy})
     end.
+
 type_check_rel_op_in(Env, ResTy, Op, P, Arg1, Arg2) ->
-    case subtype(ResTy, {type, erl_anno:new(0), 'bool', []}, Env#env.tenv) of
+    case subtype(type(boolean), ResTy, Env#env.tenv) of
         {true, Cs0} ->
           {ResTy1, VarBinds1, Cs1} = type_check_expr(Env, Arg1),
           {ResTy2, VarBinds2, Cs2} = type_check_expr(Env, Arg2),
           case compatible(ResTy1, ResTy2, Env#env.tenv) of
               {true, Cs} ->
-                  {union_var_binds([VarBinds1, VarBinds2])
+                  {union_var_binds(VarBinds1, VarBinds2, Env#env.tenv)
                   ,constraints:combine([Cs0, Cs1, Cs2, Cs])};
               false ->
                   throw({type_error, rel_error, Op, P, ResTy1, ResTy2})
           end;
         false ->
           throw({type_error, rel_error, Op, P, ResTy})
     end.
+
 type_check_list_op_in(Env, ResTy, Op, P, Arg1, Arg2) ->
-    case subtype(ResTy, {type, erl_anno:new(0), list, []}, Env#env.tenv) of
-      {true, Cs} ->
-        {VarBinds1, Cs1} = type_check_expr_in(Env, ResTy, Arg1),
-        {VarBinds2, Cs2} = type_check_expr_in(Env, ResTy, Arg2),
-          {union_var_binds([VarBinds1, VarBinds2])
-          ,constraints:combine([Cs, Cs1, Cs2])};
-      false ->
-          throw({type_error, list_op_error, Op, P, ResTy})
+    Target =
+        %% '--' always produces a proper list, but '++' gives you an improper
+        %% list if the second argument is improper.
+        case Op of
+            '--' -> type(list, [type(term)]);
+            '++' -> type(maybe_improper_list, [type(term), type(term)])
```

Comment:
There is an implementation of infimum in OTP here: https://github.com/erlang/otp/blob/master/lib/hipe/cerl/erl_types.erl#L2794 -- not the same representation of types though and maybe it is for core erlang or something, but it can perhaps serve as inspiration anyway....

Reply:
No `term()` is the right thing here. This is when computing the target type of the operator, that we then intersect with the type being pushed in. We want the maximally permissive target type, which is `[term()]` for `--` and `maybe_improper_list(term(), term())` for `++`.

Reply:
Ah, yes, I see now. Excuse my brain fart.